libs/afalg_engine/Makefile

```diff
@@ -0,0 +1,76 @@
+# Copyright (C) 2019 Eneas Ulir de Queiroz <cotequeiroz@gmail.com>
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=afalg_engine
+PKG_VERSION:=1.0.0
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://github.com/cotequeiroz/afalg_engine/archive/v$(PKG_VERSION)
+PKG_HASH:=acae83df1f69ac19944167b0a2e81960a1441a1704ff8c72ca56839aaf0dab8e
+
+PKG_MAINTAINER:=Eneas U de Queiroz <cotequeiroz@gmail.com>
+PKG_LICENSE:=Apache-2.0
+PKG_LICENSE_FILES:=LICENSE
+PKG_CONFIG_DEPENDS:= \
+	CONFIG_AFALG_ZERO_COPY \
+	CONFIG_OPENSSL_OPTIMIZE_SPEED
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/cmake.mk
+
+$(eval $(shell grep "^ENGINES_DIR" "$(TOPDIR)/package/libs/openssl/Makefile"))
+ifeq ($(ENGINES_DIR),)
+  ENGINES_DIR:=engines-1.1
+endif
+
+define Package/libopenssl-afalg_sync
+    SECTION:=libs
+    CATEGORY:=Libraries
+    SUBMENU:=SSL
+    MDEPENDS:=libopenssl
+    TITLE:=AF_ALG engine using sync crypto API
+    URL:=https://github.com/cotequeiroz/afalg_engine
+    DEPENDS:=libopenssl @OPENSSL_ENGINE @!OPENSSL_ENGINE_BUILTIN_AFALG \
+	     +libopenssl-conf +kmod-crypto-user
+    CONFLICTS:=libopenssl-afalg
+endef
+
+define Package/libopenssl-afalg_sync/description
+    This is an alternate AF_ALG engine for openssl, based on the devcrypto
+    engine, but using the AF_ALG interface instead of /dev/crypto
+
+    It is different than the AF_ALG engine that ships with OpenSSL:
+     - it uses sync calls, instead of async
+     - it suports more algorithms
+endef
+
+define Package/libopenssl-afalg_sync/config
+    config AFALG_ZERO_COPY
+        depends on PACKAGE_libopenssl-afalg_sync
+	bool "Use Zero-Copy Mode"
+	help
+	    Uses a Zero-Copy interface.  Even though it is supposed to improve
+	    performance, actual measurements indicate otherwise.
+endef
+
+ifeq ($(CONFIG_OPENSSL_OPTIMIZE_SPEED),y)
+  TARGET_CFLAGS:=$(filter-out -O%,$(TARGET_CFLAGS)) -O3
```

Comment:
Is this even necessary? -O options should have no impact on hardware acceleration speed.

Reply:
You're right, although my thought was initialization, not crypto operation.  Even though it is probably unrolling initialization loops, and initialization is slow, most time is probably spent in-kernel, not justifying the size increase.  I'll yank it out.

| |Opt Size|Opt Speed|
|-|-:|-:|
|ipk | 10,437 | 14,247|
|afalg.so|20,500|28,689|